### PR TITLE
Apache Kafka Scaler: Implementation for Excluding Persistent Lag

### PR DESCRIPTION
**Pull request by JosephABC**
### Summary
Add implementation for excluding consumer lag from partitions with persistent lag.

### Use Case
In situations where consumer is unable to process / consume from partition due to errors etc., committed offset will not change, and consumer lag on that partition will be increasing and never be decreased. KEDA trigger scaling towards the maxReplicaCount.

If partition lag is deemed as persistent, excluding its consumer lag will allow KEDA to trigger scaling appropriately based on the consumer lag observed on other topics and partition, and not be affected by this consumer lag which will not be resolved by scaling.

### Logic
Upon each polling cycle, check if current consumer offset is same as previous consumer offset.

Different: return endOffset - consumerOffset (No different from current implementation)
Same: return 0 (To exclude this partition's consumer lag from the total lag)

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [X] Tests have been added
- [X] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Relates to https://github.com/kedacore/keda/issues/3904
Relates to https://github.com/kedacore/keda-docs/pull/984

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -58,6 +58,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **General:** Add support to use pod identities for authentication in Azure Key Vault ([#3813](https://github.com/kedacore/keda/issues/3813)
 - **General:** Improve the function used to normalize metric names ([#3789](https://github.com/kedacore/keda/issues/3789)
 - **Apache Kafka Scaler:** SASL/OAuthbearer Implementation ([#3681](https://github.com/kedacore/keda/issues/3681))
+- **Apache Kafka Scaler:** SASL/OAuthbearer Implementation ([#3904](https://github.com/kedacore/keda/issues/3904))
```
Comment (JorTurFer) [inline CHANGELOG.md:61]: I guess the message isn't correct

**Review by JorTurFer — COMMENTED** [review]
I'm not really sure about this feature, in general we try to avoid maintaining information about the execution inside the scaler (`previousOffsets`). This is because scaler can be recreated when it's needed, and behaviours based on previous cycles information could work randomly.
WDYT @zroubalik ?

**Comment by zroubalik** [comment]
this test is failing:  
```
=== RUN   TestGetBrokers
    kafka_scaler_test.go:168: Expected success but got error error parsing excludePersistentLag: strconv.ParseBool: parsing "notvalid": invalid syntax
    kafka_scaler_test.go:192: Expected success but got error error parsing excludePersistentLag: strconv.ParseBool: parsing "notvalid": invalid syntax
```

**Comment by zroubalik** [comment]
> I'm not really sure about this feature, in general we try to avoid maintaining information about the execution inside the scaler (`previousOffsets`). This is because scaler can be recreated when it's needed, and behaviours based on previous cycles information could work randomly. WDYT @zroubalik ?

That's a very good point @JorTurFer, even though I **think** that it should be okay for this specific feature. @JosephABC as @JorTurFer mentioned, the scaler could be recreated during it's lifetime, would it cause any problems? I think that it will reconcile.

**Comment by zroubalik** [comment]
I'd love to see e2e test for this.

**Comment by tomkerkhove** [comment]
We are going to release KEDA v2.9 on Thursday. Do you think you can complete the open work by Wednesday @JosephABC?

**Comment by JosephABC** [comment]
@zroubalik If the scaler is recreated, I would expect the previousOffsets map to be recreated anew. So the recording of the previous offsets will start from a blank state. The downside is that the scaler will not be able to identify the partitions with persistent lag in the next reconciliation cycle and possibly trigger a scale out of the scaling target. In the following cycles, the scaler will observe the previousOffsets map and scale accordingly (excluding persistent lag if deemed necessary)

@tomkerkhove I will certainly try to complete by Wednesday. If not, I will target to complete before the next KEDA release.

**Comment by JosephABC** [comment]
@zroubalik e2e tests seems to be queued for quite a long time already. Is this normal?

**Comment by zroubalik** [comment]
> @zroubalik e2e tests seems to be queued for quite a long time already. Is this normal?

What do you mean? Only maintainers can trigger them. Do you mean this comment https://github.com/kedacore/keda/pull/3905#issuecomment-1328867804 ?

The e2e tests passed on this one (it is marked in emojis)

**Comment by JorTurFer** [comment]
> The e2e tests passed on this one (it is marked in emojis)

And you can also see it in the latest commit checks when the execution was triggered

**Comment by JosephABC** [comment]
The checks for the latest commit all passed, except for the e2e test which shows as "queued". Does that need to be runned and completed as well?

**Comment by JorTurFer** [comment]
> The checks for the latest commit all passed, except for the e2e test which shows as "queued". Does that need to be runned and completed as well?

we need to trigger them manually like here 
image

Could you add an e2e test to cover this new feature? We try to cover all the scalers with e2e tests, in this case, you could just add your feature as a test case [here](https://github.com/kedacore/keda/blob/main/tests/scalers/kafka/kafka_test.go)

File: pkg/scalers/kafka_scaler.go [hunk 2]
```diff
@@ -399,12 +413,12 @@ func (s *kafkaScaler) getConsumerOffsets(topicPartitions map[string][]int32) (*s
 	return offsets, nil
 }
 
-func (s *kafkaScaler) getLagForPartition(topic string, partitionID int32, offsets *sarama.OffsetFetchResponse, topicPartitionOffsets map[string]map[int32]int64) (int64, error) {
+func (s *kafkaScaler) getLagForPartition(topic string, partitionID int32, offsets *sarama.OffsetFetchResponse, topicPartitionOffsets map[string]map[int32]int64) (int64,  int64, error) {
```
Comment (zroubalik) [inline pkg/scalers/kafka_scaler.go:416]: Could you please document here the return values, what these represent and when these should be used? Thanks!

Comment (JosephABC) [reply pkg/scalers/kafka_scaler.go]: Added the descriptions

**Comment by zroubalik** [comment]
@tobiaskrause would you mind looking at this as well, since you are doing a Kafka PR in parallel?

**Comment by zroubalik** [comment]
@JosephABC rebase please

**Comment by tobiaskrause** [comment]
> @tobiaskrause would you mind looking at this as well, since you are doing a Kafka PR in parallel?

@zroubalik, don't see any interference with my PR

**Comment by zroubalik** [comment]
@JosephABC  could you please rebase your PR to contain only relevant commits? Thanks!

**Pull request closed without merging**